File: app/src/main/java/me/raatiniemi/worker/presentation/projects/viewmodel/CreateProjectViewModel.java
/*
 * Copyright (C) 2017 Worker Project
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package me.raatiniemi.worker.presentation.projects.viewmodel;

import android.support.annotation.NonNull;

import com.fernandocejas.frodo.annotation.RxLogObservable;

import me.raatiniemi.worker.domain.exception.InvalidProjectNameException;
import me.raatiniemi.worker.domain.exception.ProjectAlreadyExistsException;
import me.raatiniemi.worker.domain.interactor.CreateProject;
import me.raatiniemi.worker.domain.model.Project;
import me.raatiniemi.worker.domain.validator.ProjectName;
import rx.Observable;
import rx.subjects.BehaviorSubject;
import rx.subjects.PublishSubject;

import static me.raatiniemi.worker.presentation.util.RxUtil.hideErrors;
import static me.raatiniemi.worker.presentation.util.RxUtil.redirectErrors;

public interface CreateProjectViewModel {
    interface Input {
        void projectName(@NonNull String name);

        void createProject();
    }

    interface Output {
        @NonNull
        Observable<Project> createProjectSuccess();

        @NonNull
        Observable<Boolean> isProjectNameValid();
    }

    interface Error {
        @NonNull
        Observable<String> invalidProjectNameError();

        @NonNull
        Observable<String> duplicateProjectNameError();

        @NonNull
        Observable<String> createProjectError();
    }

    final class ViewModel implements Input, Output, Error {
        private final Input input;
        private final Output output;
        private final Error error;

        private final PublishSubject<String> projectName = PublishSubject.create();
        private final BehaviorSubject<Boolean> isProjectNameValid = BehaviorSubject.create(Boolean.FALSE);
        private final PublishSubject<Void> createProject = PublishSubject.create();
        private final PublishSubject<Project> createProjectSuccess = PublishSubject.create();
        private final PublishSubject<Throwable> createProjectError = PublishSubject.create();

        private final CreateProject useCase;

        public ViewModel(@NonNull CreateProject useCase) {
            input = this;
            output = this;
            error = this;

            this.useCase = useCase;

            projectName.map(ProjectName::isValid)
                    .subscribe(isProjectNameValid);

            createProject.withLatestFrom(projectName, (event, name) -> name)
                    .switchMap(name -> executeUseCase(name)
                            .compose(redirectErrors(createProjectError))
                            .compose(hideErrors()))
                    .subscribe(createProjectSuccess);
        }

        @NonNull
        private Observable<Project> executeUseCase(@NonNull String name) {
            try {
                Project project = Project.builder(name)
                        .build();

                return Observable.just(useCase.execute(project));
            } catch (Exception e) {
                return Observable.error(e);
            }
        }

        @Override
        public void projectName(@NonNull String name) {
            projectName.onNext(name);
        }

        @Override
        public void createProject() {
            createProject.onNext(null);
        }

        @NonNull
        @Override
        @RxLogObservable
        public Observable<Boolean> isProjectNameValid() {
            return isProjectNameValid;
        }

        @NonNull
        @Override
        @RxLogObservable
        public Observable<Project> createProjectSuccess() {
            return createProjectSuccess.asObservable();
        }

        @NonNull
        @Override
        @RxLogObservable
        public Observable<String> invalidProjectNameError() {
            return createProjectError
                    .filter(ViewModel::isInvalidProjectNameError)
                    .map(Throwable::getMessage);
        }

        private static boolean isInvalidProjectNameError(@NonNull Throwable e) {
            return e instanceof InvalidProjectNameException;
        }

        @NonNull
        @Override
        @RxLogObservable
        public Observable<String> duplicateProjectNameError() {
            return createProjectError
                    .filter(ViewModel::isDuplicateProjectNameError)
                    .map(Throwable::getMessage);
        }

        private static boolean isDuplicateProjectNameError(@NonNull Throwable e) {
            return e instanceof ProjectAlreadyExistsException;
        }

        @NonNull
        @Override
        @RxLogObservable
        public Observable<String> createProjectError() {
            return createProjectError
                    .filter(ViewModel::isUnknownError)
                    .map(Throwable::getMessage);
        }

        private static boolean isUnknownError(@NonNull Throwable e) {
            return !isInvalidProjectNameError(e) && !isDuplicateProjectNameError(e);
        }

        @NonNull
        public Input input() {
            return input;
        }

        @NonNull
        public Output output() {
            return output;
        }

        @NonNull
        public Error error() {
            return error;
        }
    }
}

File: app/src/main/java/me/raatiniemi/worker/feature/projects/createproject/viewmodel/CreateProjectViewModel.kt
/*
 * Copyright (C) 2021 Tobias Raatiniemi
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package me.raatiniemi.worker.feature.projects.createproject.viewmodel

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import me.raatiniemi.worker.domain.project.model.projectName
import me.raatiniemi.worker.domain.project.usecase.CreateProject
import me.raatiniemi.worker.domain.project.usecase.FindProject
import me.raatiniemi.worker.domain.project.usecase.InvalidProjectNameException
import me.raatiniemi.worker.domain.project.usecase.ProjectAlreadyExistsException
import me.raatiniemi.worker.feature.projects.createproject.model.CreateProjectError
import me.raatiniemi.worker.feature.projects.createproject.model.CreateProjectViewActions
import me.raatiniemi.worker.feature.shared.model.ConsumableLiveData
import me.raatiniemi.worker.feature.shared.model.Error
import me.raatiniemi.worker.feature.shared.model.plusAssign
import me.raatiniemi.worker.monitor.analytics.Event
import me.raatiniemi.worker.monitor.analytics.UsageAnalytics
import timber.log.Timber

internal class CreateProjectViewModel(
    private val usageAnalytics: UsageAnalytics,
    private val createProject: CreateProject,
    private val findProject: FindProject
) : ViewModel() {
    private val _name = MutableLiveData<String>()
    var name: LiveData<String> = _name

    private val _error = MutableLiveData<Error?>()
    val error: LiveData<Error?> = _error

    val viewActions = ConsumableLiveData<CreateProjectViewActions>()

    suspend fun onNameChange(name: String) {
        _error += null
        _name += name

        try {
            val project = findProject(projectName(name))
            if (project != null) {
                _error += CreateProjectError.ProjectAlreadyExists
            }
        } catch (e: InvalidProjectNameException) {
            _error += CreateProjectError.InvalidName
        }
    }

    suspend fun createProject(name: String) {
        try {
            val project = createProject(projectName(name))

            usageAnalytics.log(Event.ProjectCreate)
            viewActions += CreateProjectViewActions.Created(project)
        } catch (e: ProjectAlreadyExistsException) {
            Timber.d("Project with name \"$name\" already exists")
            _error += CreateProjectError.ProjectAlreadyExists
        } catch (e: InvalidProjectNameException) {
            Timber.w("Project name \"$name\" is not valid")
            _error += CreateProjectError.InvalidName
        } catch (e: Exception) {
            Timber.w(e, "Unable to create project")
            _error += CreateProjectError.Unknown
        }
    }

    fun dismiss() {
        viewActions += CreateProjectViewActions.Dismiss
    }
}
